File: programs/dirent/examples/deepdir.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>

void list_directory(const char *dirname, int depth) {
    DIR *dir = opendir(dirname);
    if (!dir) {
        perror("opendir");
        return;
    }

    struct dirent *entry;
    while ((entry = readdir(dir)) != NULL) {
        if (strcmp(entry->d_name, ".") == 0 || strcmp(entry->d_name, "..") == 0)
            continue;

        printf("%*s%s\n", depth * 2, "", entry->d_name);

        if (entry->d_type == DT_DIR) {
            char path[1024];
            snprintf(path, sizeof(path), "%s/%s", dirname, entry->d_name);
            list_directory(path, depth + 1);
        }
    }

    closedir(dir);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <directory>\n", argv[0]);
        return EXIT_FAILURE;
    }

    list_directory(argv[1], 0);
    return EXIT_SUCCESS;
}

File: programs/dirent/examples/rust/deepdir.rs
use std::fs;

fn list_directory(dir: &str, depth: usize) -> std::io::Result<()> {
    for entry in fs::read_dir(dir)? {
        let entry = entry?;
        let file_name = entry.file_name();
        let file_name_str = file_name.to_string_lossy();

        println!("{:width$}{}", "", file_name_str, width = depth * 2);

        if entry.file_type()?.is_dir() {
            let path = entry.path();
            list_directory(path.to_str().unwrap(), depth + 1)?;
        }
    }
    Ok(())
}

fn main() -> std::io::Result<()> {
    let args: Vec<String> = std::env::args().collect();

    if args.len() != 2 {
        eprintln!("Usage: {} <directory>", &args[0]);
        std::process::exit(1);
    }

    list_directory(&args[1], 0)?;
    Ok(())
}
